ykllvmwrap/src/jitmodbuilder.cc

```diff
@@ -979,8 +979,15 @@ class JITModBuilder {
           continue;
 
         if (isa<CallInst>(I)) {
-
           if (isa<IntrinsicInst>(I)) {
+            // llvm.lifetime.start.* and llvm.lifetime.end.* have no meaninful
+            // program semantics for our purposes, and generate no code anyway.
```

Comment:
"For our purposes"?

Reply:
Calls to these functions are only used by LLVM's stack slot colouring pass.

Reply:
I plead ignorance: are these annotations not needed at all, or not necessary by our stage in the pipeline? And how do these affect recursion?

Reply:
Sorry, let me elaborate.

Calls to `llvm.lifetime.*` are annotations added by upstream LLVM at some stage in their AOT pipeline for a later optimisation pass (of theirs) to pick up. For whatever reason, they leave the annotations in, and thus we see them in our AOT module.

Currently, when the trace compiler sees a call to an intrinsic, it needs to know a) was the call inlined? and if not, b) what symbol does it need to call then.

For `llvm.lifetime.*` there is no symbol that it could call and there is no code that could have been inlined, as these calls are not really calls. They generate no code and have no program semantics.

So all I'm doing here is teaching the trace compiler to ignore them.

I hope that makes sense?

Reply:
Should we filter these out at the AOT stage?

Reply:
We could but:

 - I don't know if there are runtime analyses in the wild that rely on these annotations in embedded IR.
 - I don't know if there are LLVM plugin passes (external to the LLVM codebase) that rely on these annotations.
 - It's harder for us to do, as we'd have to add a whole pass.

Reply:
Fair enough. Can we at least clarify more thoroughly in this PR *why* we're excluding these? [As an aside: presumably we expect LLVM to add these back in while compiling the trace if it needs to?]

Reply:
> presumably we expect LLVM to add these back in while compiling the trace if it needs to?

That's my understanding.

Reply:
@bjorn3 Is this only used by rustc (and not any other front end)?

Reply:
It is valid to omit them entirely, but doing so will increase stack usage.

Reply:
OK, if we update the comment, I think we'll be good to move on.

Reply:
Try [0264dc7](https://github.com/ykjit/yk/pull/528/commits/0264dc7b9be43d738c798ffa8c1c86d4c99ae8d4)